src/vdom/component.js

```diff
@@ -105,40 +118,49 @@ export function renderComponent(component, opts, mountAll, isChild) {
 		let childComponent = rendered && rendered.nodeName,
 			toUnmount, base;
 
-		if (typeof childComponent==='function') {
-			// set up high order component link
+		try {
+			if (typeof childComponent==='function') {
+				// set up high order component link
 
-			let childProps = getNodeProps(rendered);
-			inst = initialChildComponent;
+				let childProps = getNodeProps(rendered);
+				inst = initialChildComponent;
 
-			if (inst && inst.constructor===childComponent && childProps.key==inst.__key) {
-				setComponentProps(inst, childProps, SYNC_RENDER, context, false);
-			}
-			else {
-				toUnmount = inst;
+				if (inst && inst.constructor===childComponent && childProps.key==inst.__key) {
+					setComponentProps(inst, childProps, SYNC_RENDER, context, false);
+				}
+				else {
+					toUnmount = inst;
+
+					component._component = inst = createComponent(childComponent, childProps, context, component);
+					inst.nextBase = inst.nextBase || nextBase;
+					inst._parentComponent = component;
+					setComponentProps(inst, childProps, NO_RENDER, context, false);
+					renderComponent(inst, SYNC_RENDER, mountAll, true);
+				}
 
-				component._component = inst = createComponent(childComponent, childProps, context);
-				inst.nextBase = inst.nextBase || nextBase;
-				inst._parentComponent = component;
-				setComponentProps(inst, childProps, NO_RENDER, context, false);
-				renderComponent(inst, SYNC_RENDER, mountAll, true);
+				base = inst.base;
 			}
+			else {
+				cbase = initialBase;
 
-			base = inst.base;
-		}
-		else {
-			cbase = initialBase;
+				// destroy high order component link
+				toUnmount = initialChildComponent;
+				if (toUnmount) {
+					cbase = component._component = null;
+				}
 
-			// destroy high order component link
-			toUnmount = initialChildComponent;
-			if (toUnmount) {
-				cbase = component._component = null;
+				if (initialBase || (opts&SYNC_RENDER)) {
+					if (cbase) cbase._component = null;
+					base = diff(cbase, rendered, context, mountAll || !isUpdate, initialBase && initialBase.parentNode, component);
+				}
 			}
-
-			if (initialBase || opts===SYNC_RENDER) {
-				if (cbase) cbase._component = null;
-				base = diff(cbase, rendered, context, mountAll || !isUpdate, initialBase && initialBase.parentNode, true);
+		} catch (e) {
+			if (opts == REATTEMPT_RENDER) {
+				throw e;
 			}
+			base = initialBase || document.createTextNode("");
+			exception = e;
+			hasException = true;
```

Comment:
The `catchErrorInComponent` would live here (directly). If that function doesn't / can't unmount the tree directly (not sure), then `hasException` can be renamed to `toDismount` & trigger the flush + re-render below.